pkg/jobrunaggregator/jobrunaggregatorlib/util.go

```diff
@@ -0,0 +1,149 @@
+package jobrunaggregatorlib
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+
+	"github.com/openshift/ci-tools/pkg/jobrunaggregator/jobrunaggregatorapi"
+)
+
+type JobRunGetter interface {
+	// GetRelatedJobRuns gets all related job runs for analysis
+	GetRelatedJobRuns(ctx context.Context) ([]jobrunaggregatorapi.JobRunInfo, error)
+}
+
+// WaitForAllJobRuns waits for all runs until they finish or timeToStopWaiting
+func WaitForAllJobRuns(ctx context.Context,
+	readyAt time.Time,
+	timeToStopWaiting time.Time,
+	jobRunGetter JobRunGetter) error {
+	clock := clock.RealClock{}
+	fmt.Printf("Wait and get job runs. now=%v, ReadyAt=%v, timeToStopWaiting=%v.\n", clock.Now(), readyAt, timeToStopWaiting)
+
+	var unfinishedJobRunNames []string
+	for {
+		fmt.Println() // for prettier logs
+		unfinishedJobRunNames = []string{}
+
+		if clock.Now().Before(readyAt) {
+			fmt.Printf("waiting until %v to collect more jobRuns before assessing finished or not. (now=%v)\n", readyAt, clock.Now())
+			time.Sleep(2 * time.Minute)
+			continue
+		}
+		fmt.Printf("it is %v, finished waiting until %v.\n", clock.Now(), readyAt)
```

Comment:
this is the initial wait.  So you have this wait as one method, then the rest as the get all jobs.

Reply:
In my current change, all wait (ReadyAt and timeToStopWaiting) are both in the wait function. By the time this function returns, the wait is over, either all jobs done or the timeToStopWaiting reached. Did you mean to have something different?